### PR TITLE
Improve testharness "function throws" assertions.

### LINKED ISSUE #11726: assert_throws is too loose, misses browser bugs
I discovered today that some browsers throw a `DOMException` with `name == "TypeError"` in some cases when specs say to throw a `TypeError`.

Unfortunately, wpt's `assert_throws(new TypeError(), stuff)` doesn't catch this, because it just compares `.name` values.

I was going to suggest comparing `.constructor.name` values as well, but then I discovered https://github.com/web-platform-tests/wpt/pull/11643 and the issues mentioned in the review comments there.

It seems like we really want the following primitives for testing throwing behavior:

1. Test that a DOMException is thrown with the given name/code and corresponding code/name.  Right now you can sort of get that by passing a string name or integer code as the first arg, instead of an object, though even then `assert_throws` doesn't check that the result is a `DOMException` in any meaningful way (e.g. by checking `constructor.name` or trying to use a `DOMException.prototype` getter on it or something).

2. Test that an exception is thrown with the given `.constructor.name`, or whatever other way we can think of to detect a "real TypeError".

3.  Test whatever https://github.com/web-platform-tests/wpt/pull/11643 is trying to test.

The current `assert_throws` does a mix of these, without getting any of them right.

Are there other things one might want to test?  Should we even be trying to shoehorn all of these into a single `assert_throws`?

//cc @zcorpan @saschanaz @gsnedders @jgraham @annevk @domenic

### PR DESCRIPTION
**Pull request by bzbarsky**
We add new functions to allow more precise testing of what sorts of
exceptions get thrown.

Fixes https://github.com/web-platform-tests/wpt/issues/11726

### CODE REVIEW

**Review by Ms2ger — CHANGES_REQUESTED** [review]
Please wait for https://github.com/web-platform-tests/rfcs/pull/29

File: resources/testharness.js [hunk 1]
```diff
@@ -1482,6 +1503,246 @@ policies and contribution forms [3].
...
+     *
+     * @param {object} constructor The expected exception constructor.
+     * @param {Function} func Function which should throw.
+     * @param {string} description Error description for the case that the error is not thrown.
+     */
+    function assert_throws_js(constructor, func, description)
+    {
+        try {
+            func.call(this);
+            assert(false, "assert_throws_js", description,
```
[the remark this answers is not in the document]
Comment (bzbarsky) [reply resources/testharness.js]: Um... I did add those, no?  See `promise_rejects_js` and so forth...

[the remark this answers is not in the document]
Comment (bzbarsky) [reply resources/testharness.js]: Ah, thank you for explaining offline what you meant.  Updated.

**Comment by jugglinmike** [comment]
Could you also add tests for the new functions? The test suite for testharness.js is in [the `resources/test` directory](https://github.com/web-platform-tests/wpt/tree/master/resources/test). (Although we originally implemented the framework tests as so-called "integration tests," experience has shown that format to be a little awkward. [This test for "exceptional
cases"](https://github.com/web-platform-tests/wpt/blob/4ca7001dcf130f8dad3d831acf0336d889e5c49a/resources/test/tests/unit/exceptional-cases.html) demonstrates a more ergonomic approach.)

**Comment by bzbarsky** [comment]
I can do that.  Unfortunately, it seems like I can't test "this assertion should _fail_ in this case", as far as I can tell: all assertions in the test have to pass...

**Comment by bzbarsky** [comment]
> Could you also add tests for the new functions? 

Thank you.  Of course those showed a bug in the code... ;)  Thanks for the pointer to the test setup as well.  Fixed code and tests coming up.

**Comment by jugglinmike** [comment]
Tests are magic that way :)

**Comment by jgraham** [comment]
So I just merged the RFC. It seems like using `same_value` rather than `Object.is` is better at this time; I'm hoping with that change @Ms2ger will do another round of review and everything will be fine (@Ms2ger if you aren't able to do that review please let me know).

**Comment by bzbarsky** [comment]
Updated to `same_value`.

**Review by Ms2ger — APPROVED** [review]
lgtm

**Comment by domenic** [comment]
I'm so excited for this. /cc @ricea; this can clean up the streams tests a bit. Thanks @bzbarsky for driving this from idea to execution.

**Comment by jdm** [comment]
Was this change expected to update https://web-platform-tests.org/writing-tests/testharness-api.html or do the docs need a separate change to update them?

**Comment by bzbarsky** [comment]
@jdm That needs a separate change.  I'll try to get something put together if someone doesn't beat me to it.

**Comment by bzbarsky** [comment]
> Was this change expected to update https://web-platform-tests.org/writing-tests/testharness-api.html

To close the loop, this got updated in https://github.com/web-platform-tests/wpt/pull/21762

**Pull request merged**